### PR TITLE
Optimize sample and list operations, including auditing and user lookups

### PR DESCRIPTION
**Pull request by labkey-jeckels**
## Rationale
Speed up sample and list operations by taking per-row work out of the audit write path, the list audit formatter, and import lookup resolution. Create custom sample type indices in the materialized tables so that grids over large sample types stop scanning.

## Related Pull Requests
- https://github.com/LabKey/platform/pull/8020
- https://github.com/LabKey/limsModules/pull/2461

## Changes
- Collect audit events raised while building another audit record and write them in batches rather than one insert each                                                                                                                  
- Cache each audit type's provisioned storage table instead of rebuilding a schema and table for every event inserted                                                                                                                    
- Resolve the list audit formatter once per operation instead of once per row, and batch the row events it produces                                                                                                                      
- Stop discarding import lookup memoization mid import so a repeated value resolves once                                                                                                                                                 
- Mirror a sample type's admin defined indices onto its materialized table, and add a name index                                                                                                                                         
- Build the materialized table's non essential indices after it is published to readers so queries are not held behind the slowest index                                                                                                 
- Scope cached transaction audit events by container and audit log permission so they match what a database read returns                                                                                                                 
- Add integration tests for the list audit batch boundary, the mirrored domain index, lookup memoization, and cached transaction event scoping     

## Tasks
- [ ] Claude Code Review
- [ ] Manual Testing
- [x] Test Automation

### CODE REVIEW

File: audit/src/org/labkey/audit/AuditLogImpl.java [hunk 1]
```diff
@@ -248,6 +269,47 @@ public ActionURL getAuditUrl()
...
+     * Mirrors {@link ContainerFilter}'s SQL: an event matches on its own container, or on being a child of an in-scope
+     * container whose type the filter includes. Cached events need this applied by hand -- reading them back from the
+     * database is what would otherwise apply it.
+     */
+    private static Predicate<AuditTypeEvent> inScope(User user, Container container, @Nullable ContainerFilter containerFilter)
+    {
+        ContainerFilter cf = null == containerFilter ? ContainerFilter.current(container, user) : containerFilter;
+        Collection<GUID> ids = scopeIds(user, container, cf);
+        if (null == ids)
+            return event -> true;
```
Comment (labkey-nicka) [inline audit/src/org/labkey/audit/AuditLogImpl.java]: nit: Rename `event` to `_`.

File: audit/src/org/labkey/audit/AuditLogImpl.java [hunk 2]
```diff
@@ -248,6 +269,47 @@ public ActionURL getAuditUrl()
...
+
+        Set<GUID> scope = new HashSet<>(ids);
```
Comment (labkey-nicka) [inline audit/src/org/labkey/audit/AuditLogImpl.java:285]: Just curious, why copy the collection here? To ensure it is a set?

File: experiment/src/org/labkey/experiment/api/ExpMaterialTableImpl.java [hunk 3]
```diff
@@ -1376,6 +1386,75 @@ private _MaterializedQueryHelper getOrCreateMQH()
...
+            // Ordinal names because ${NAME} is already 33 of the 63 chars Postgres allows before it silently truncates and collides.
+            ddl.add("CREATE INDEX idx_${NAME}_d" + ddl.size() + " ON temp.${NAME} (" + String.join(", ", identifiers) + ")");
+        }
+
+        return ddl;
+    }
+
+    private static String indexKey(List<String> columnNames)
+    {
+        return columnNames.stream().map(name -> name.toLowerCase(Locale.ROOT)).collect(Collectors.joining(","));
```
Comment (labkey-nicka) [inline experiment/src/org/labkey/experiment/api/ExpMaterialTableImpl.java:1455]: Should this sort the columns so that the key is consistent/reproducible?

**Review by labkey-nicka — APPROVED** [review]

File: experiment/src/org/labkey/experiment/api/ExpMaterialTableImpl.java (hunk 3 above)
Comment (labkey-jeckels) [reply experiment/src/org/labkey/experiment/api/ExpMaterialTableImpl.java:1455]: I don't think so. An index over `(x,y)` is not the same as one over `(y,x)`. The UI doesn't let you define custom multi-column indices today but the API does.

File: audit/src/org/labkey/audit/AuditLogImpl.java (hunk 2 above)
Comment (labkey-jeckels) [reply audit/src/org/labkey/audit/AuditLogImpl.java:285]: It's more important to have a hashed collection than to ensure no dupes. Realistically, this won't be huge in practice, but better to guarantee an `O(1)` lookup. We should probably swap from `Collection` to `Set` in `ContainerFilter` but that's way out of scope.

File: api/src/org/labkey/api/audit/AbstractAuditHandler.java [hunk 4]
```diff
@@ -26,14 +26,19 @@
...
 
 import static org.labkey.api.gwt.client.AuditBehaviorType.SUMMARY;
 
 public abstract class AbstractAuditHandler implements AuditHandler
 {
-    protected abstract AuditTypeEvent createSummaryAuditRecord(User user, Container c, AuditConfigurable tInfo, QueryService.AuditAction action, @Nullable String userComment, int rowCount, @Nullable Map<String, Object> row);
+    /** Bounds both the JDBC batch and the events held in memory while a batch accumulates. */
+    public static final int AUDIT_BATCH_SIZE = 2500;
+
+    protected abstract AuditTypeEvent createSummaryAuditRecord(User user, Container c, AuditConfigurable tInfo, QueryService.AuditAction action, @Nullable String userComment, int rowCount, @Nullable Map<String, Object> row, List<AuditTypeEvent> sideEffectEvents);
```
Comment (XingY) [inline api/src/org/labkey/api/audit/AbstractAuditHandler.java:41]: Like mentioned in the other PR, I think we can skip createSummaryAuditRecord changes.

**Review by XingY — APPROVED** [review]
